File: backend/hotel-main/src/main/java/at/fhtw/hotel/mapper/RoomMapper.java
package at.fhtw.hotel.mapper;

import at.fhtw.hotel.model.Extra;
import at.fhtw.hotel.model.Room;
import at.fhtw.hotel.model.RoomImage;
import at.fhtw.hotel.persistence.entity.ExtraEntity;
import at.fhtw.hotel.persistence.entity.RoomEntity;
import at.fhtw.hotel.persistence.entity.RoomImageEntity;
import at.fhtw.hotel.persistence.entity.RoomExtraEntity;
import java.util.Comparator;
import java.util.List;

public class RoomMapper {

    public Room toDomain(RoomEntity entity) {
        List<RoomImage> images = entity.getImages().stream()
                .sorted(Comparator.comparingInt(RoomImageEntity::getSortOrder))
                .map(this::toDomain)
                .toList();

        List<Extra> extras = entity.getRoomExtras().stream()
                .map(RoomExtraEntity::getExtra)
                .map(this::toDomain)
                .toList();

        return Room.builder()
                .id(entity.getId())
                .title(entity.getTitle())
                .description(entity.getDescription())
                .maxGuests(entity.getMaxGuests())
                .basePricePerNight(entity.getBasePricePerNight())
                .images(images)
                .extras(extras)
                .build();
    }

    private RoomImage toDomain(RoomImageEntity entity) {
        return RoomImage.builder()
                .id(entity.getId())
                .roomId(entity.getRoom().getId())
                .fileName(entity.getFileName())
                .altText(entity.getAltText())
                .sortOrder(entity.getSortOrder())
                .build();
    }

    private Extra toDomain(ExtraEntity entity) {
        return Extra.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .title(entity.getTitle())
                .description(entity.getDescription())
                .iconName(entity.getIconName())
                .build();
    }
}

File: frontend/src/infrastructure/mappers/roomMapper.ts
import type { Extra, Room, RoomImage } from "../../core/models/room";

export interface RoomImageApi {
  id: number;
  url: string;
  sort_order: number;
  alt_text: string;
}

export interface ExtraApi {
  id: number;
  code: string;
  title: string;
  description: string;
  icon_name: string;
}

export interface RoomApi {
  id: number;
  title: string;
  description: string;
  max_guests: number;
  base_price_per_night: number;
  images: RoomImageApi[];
  extras: ExtraApi[];
}

export const mapRoomImage = (image: RoomImageApi): RoomImage => ({
  id: image.id,
  url: image.url,
  sortOrder: image.sort_order,
  altText: image.alt_text,
});

export const mapExtra = (extra: ExtraApi): Extra => ({
  id: extra.id,
  code: extra.code,
  title: extra.title,
  description: extra.description,
  iconName: extra.icon_name,
});

export const mapRoom = (room: RoomApi): Room => ({
  id: room.id,
  title: room.title,
  description: room.description,
  maxGuests: room.max_guests,
  basePricePerNight: room.base_price_per_night,
  images: room.images.map(mapRoomImage),
  extras: room.extras.map(mapExtra),
});
